{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Demo on DLPFC tissue\n",
    "We use of section 151509 of the human dorsolateral prefrontal cortex (DLPFC) tissue as an example to show how to use srstGCN to obtain super-resolution gene expression profiles by weakly supervised GCN framework. srstGCN enhances gene expression to single-cell resolution and can also predict gene expression both outside the spots and in external tissue sections at the single-cell level. The 10X Visium data of section 151509 can be downloaded from"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Data preparation"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-07-15T14:16:38.517641100Z",
     "start_time": "2024-07-15T14:15:52.594299300Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image loaded from E:\\超分项目\\srstGCN\\data\\he-raw.jpg\n",
      "E:\\超分项目\\srstGCN\\data\\he.jpg\n",
      "Dataframe loaded from E:\\超分项目\\srstGCN\\data\\locs-raw.csv\n"
     ]
    }
   ],
   "source": [
    "import rescale\n",
    "import os\n",
    "prefix =f'{os.getcwd()}\\\\data\\\\'\n",
    "# rescale.main(prefix, image=True, locs=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Mutimodel feature map extraction"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Token has not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.\n",
      "Token is valid (permission: read).\n",
      "Your token has been saved to C:\\Users\\DELL\\.cache\\huggingface\\token\n",
      "Login successful\n",
      "The size of the histological image is:(19488, 19488, 3)\n",
      "Image loaded from E:\\超分项目\\srstGCN\\data\\he.jpg\n"
     ]
    }
   ],
   "source": [
    "import feature_map_extraction as srstGCN\n",
    "import pickle\n",
    "login = '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'\n",
    "img_emb = srstGCN.his_ex(prefix, login)\n",
    "rgb_emb = srstGCN.rgb_ex(prefix)\n",
    "pos_emb = srstGCN.pos_ex(prefix, img_emb)\n",
    "embs = dict(his=img_emb, rgb=rgb_emb, pos=pos_emb)\n",
    "with open(prefix + 'embeddings-hist.pickle', 'wb') as file:\n",
    "    pickle.dump(embs, file)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-07-15T14:20:54.980629400Z",
     "start_time": "2024-07-15T14:17:42.609205500Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Select highly variable genes and generate mask image"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pickle loaded from E:\\超分项目\\srstGCN\\data\\embeddings-hist.pickle\n",
      "Clustering pixels using km...\n",
      "(1406596, 1029)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\DELL\\anaconda3\\envs\\pytorch\\Lib\\site-packages\\sklearn\\cluster\\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  super()._check_params_vs_input(X, default_n_init=10)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_clusters: 2\n",
      "E:\\超分项目\\srstGCN\\data\\mask-small.png\n",
      "Dataframe loaded from E:\\超分项目\\srstGCN\\data\\cnts.csv\n",
      "E:\\超分项目\\srstGCN\\data\\gene-names.txt\n"
     ]
    }
   ],
   "source": [
    "import get_mask\n",
    "import select_genes\n",
    "get_mask.main(prefix)\n",
    "select_genes.main(prefix, n_top=1000)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-07-15T14:26:16.741301200Z",
     "start_time": "2024-07-15T14:23:20.226051100Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Inferring sing-cell resolution gene expression after the train process"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from impute import main as train_model\n",
    "train_model(prefix, epoch=500, device='cuda', n_states=5, load_saved=False)"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
